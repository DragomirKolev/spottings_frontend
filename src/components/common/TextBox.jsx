import React, { useState } from 'react';
import { createUseStyles } from 'react-jss';

import { GLOBAL_SPACING, styles } from '../../styles/common';

const useStyles = createUseStyles({
  container: {
    display: 'flex',
    justifyContent: 'flex-end',
    borderBottom: styles.border('green'),
    paddingBottom: GLOBAL_SPACING * 2,
    marginTop: GLOBAL_SPACING,
    marginBottom: GLOBAL_SPACING,
  },
  inputContainer: {
    width: '100%',
    marginRight: GLOBAL_SPACING * 4,
  },
  label: { marginRight: GLOBAL_SPACING * 2, width: '25%' },
  box: styles.textBox,
  greenBox: { border: styles.border('green'), ...styles.textBox },
  redBox: { border: styles.border('red'), ...styles.textBox },
  dropdown: {
    display: 'block',
    height: ({ search }) => search && Math.min(search.length * 20, 60),
    overflow: 'auto',
    backgroundColor: 'white',
  },
  searchLabel: {
    display: 'block',
    borderBottom: styles.border('green'),
    cursor: 'pointer'
  },
  helpLabel: {
    color: 'red',
  }
});

const TextBox = ({ label, value, help, onChange, search = null }) => {
  const classes = useStyles({ search });
  const [searchOpen, setSearchOpen] = useState(false);

  const className = help && !value ? classes.redBox 
    : value ? classes.greenBox : classes.box;

  const searchMap = search && search.map((label, idx) =>
    <div key={idx} onClick={() => {
      onChange(label);
      setSearchOpen(false)
    }} 
    className={classes.searchLabel}>{label}</div>
  );

  return (
    <div className={classes.container}>
      <div className={classes.label}>{label} :</div>
      <div className={classes.inputContainer}>
        <input
          className={className}
          type='text'
          placeholder='Type here...'
          value={value || ''}
          onFocus={() => setSearchOpen(true)}
          onChange={(evt) => onChange(evt.currentTarget.value)} 
        />
       {className === classes.redBox && <div className={classes.helpLabel}>Please fill in this field.</div>}
       {searchOpen && search && <div className={classes.dropdown}>{searchMap}</div>}
      </div>
    </div>
  );
}

export default TextBox;