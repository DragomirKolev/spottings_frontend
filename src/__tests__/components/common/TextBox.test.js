import React from 'react';
import renderer from 'react-test-renderer';

import TextBox from '../../../components/common/TextBox';

it('renders TextBox', () => {
  const tree = renderer
    .create(<TextBox 
          label={'Confident in your sighting?'}
          value='testValue'
          search={['te', 'tes', 'test']}
          help={false}
          onChange={jest.fn()}
      />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});