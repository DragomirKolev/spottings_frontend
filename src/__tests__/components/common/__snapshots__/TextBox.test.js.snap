// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders TextBox 1`] = `
<div
  className="container-0-2-1"
>
  <div
    className="label-0-2-3"
  >
    Confident in your sighting?
     :
  </div>
  <div
    className="inputContainer-0-2-2"
  >
    <input
      className="greenBox-0-2-5"
      onChange={[Function]}
      onFocus={[Function]}
      placeholder="Type here..."
      type="text"
      value="testValue"
    />
  </div>
</div>
`;
